### PR TITLE
fix: spec compliant parseInt

### PR DESCRIPTION
**Pull request by MaxGraey**
- more rejection tests
- more tests which mixed different prefixes and radices
- fixes edge cases for `parseInt`
- proper generic variants of `strtol` for `F64.parseInt` and `F32.parseInt`

- [x] I've read the contributing guidelines

### CODE REVIEW

File: std/assembly/util/string.ts [hunk 1]
```diff
@@ -652,39 +652,48 @@ export function toUpper8(c: u32): u32 {
...
+    }
   }
 
   var ptr = changetype<usize>(str) /* + HEAD -> offset */;
   var code = <u32>load<u16>(ptr);
 
   // determine sign
   // @ts-ignore: cast
-  var sign: T = 1;
+  var sign = false;
```
Comment (dcodeIO) [inline std/assembly/util/string.ts:669]: Does this fix something? If so, then there are a couple unnecessary `ts-ignore`s now besides the branch it adds.

**Comment by dcodeIO** [comment]
Iiuc, on Discord it was also reported that global `parseInt` returns `0` instead of `NaN` for inputs like `"this is not semantically equal to 0"`. Does this PR fix this as well, does not attempt to fix it, or was this a false alarm?

Example given was

```ts
assert(parseInt("this is not semantically equal to 0") == 0)
assert(parseInt("this is not semantically equal to 0") == parseInt("0"))
```

mentioning that these asserts pass, which seems weird given that global `parseInt` is

```ts
export function parseInt(str: string, radix: i32 = 0): f64 {
  return strtol<f64>(str, radix);
}
```

**Comment by dcodeIO** [comment]
From a quick test it seems that

```ts
parseInt("") // returns NaN as expected but
parseInt("abc") // returns 0, which is strange
```

**Comment by MaxGraey** [comment]
In my tests `parseInt("abc")` equal to `NaN` but not a zero.

**Comment by MaxGraey** [comment]
> Iiuc, on Discord it was also reported that global parseInt returns 0 instead of NaN for inputs like "this is not semantically equal to 0". Does this PR fix this as well, does not attempt to fix it, or was this a false alarm?

It fix already halfbacked solution which just not properly tests. So as you suggested in discord solution:
```ts
let parsedFloat = parseInt('bad input');
if(isNaN(parsedFloat)) {
   // fail
} else {
  // success
}
```
Just not worked. Current PR fix this.

**Comment by dcodeIO** [comment]
Great, thanks, looks good then with the other comment above addressed :)

**Comment by MaxGraey** [comment]
Just need one edge case. But it will be tomorrow, so plz don't merge it yet

**Comment by MaxGraey** [comment]
Alright. It seems all cases handled properly

**Comment by MaxGraey** [comment]
@dcodeIO I have some race logic which don't know how better handle:
```ts
assert(parseInt("0b11", 16) == 0xb11);
assert(parseInt("0x11", 16) == 0x11);
```
Both should valid. But I'm struggling how this properly handle)

File: std/assembly/number.ts [hunk 2]
```diff
@@ -295,7 +295,7 @@ export abstract class F32 {
   }
 
   static parseInt(value: string, radix: i32 = 0): f32 {
-    return <f32>strtol<i64>(value, radix);
+    return <f32>strtol<f64>(value, radix);
```
Comment (dcodeIO) [inline std/assembly/number.ts:298]: Does `strtol<f32>` suffer from precision issues here?

Comment (MaxGraey) [reply std/assembly/number.ts:298]: yes

File: std/assembly/util/string.ts [hunk 3]
```diff
@@ -652,45 +652,66 @@ export function toUpper8(c: u32): u32 {
...
+        // @ts-ignore: cast
+        return <T>0;
+      }
+    }
+    // handle case as parseInt("0xFF", 16) by spec
+    if (radix == 16) {
+      if (
+        len > 2 &&
+        code == CharCode._0 &&
+        (<u32>load<u16>(ptr + 2) | 32) == CharCode.x
```
Comment (dcodeIO) [inline std/assembly/util/string.ts:708]:
```suggestion
        (<u32>load<u16>(ptr, 2) | 32) == CharCode.x
```

File: std/assembly/util/string.ts [hunk 4]
```diff
@@ -652,45 +652,66 @@ export function toUpper8(c: u32): u32 {
...
+      ) {
+        ptr += 4; len -= 2;
+      }
+    }
+  } else {
+    // determine radix by literal prefix
     if (code == CharCode._0 && len > 2) {
       switch (<u32>load<u16>(ptr + 2) | 32) {
```
Comment (dcodeIO) [inline std/assembly/util/string.ts:716]: Not part of this PR, but perhaps also offset=2

File: std/assembly/util/string.ts [hunk 5]
```diff
@@ -770,7 +797,7 @@ export function strtod(str: string): f64 {
       load<u64>(ptr, 0) == 0x690066006E0049 && // ifnI
       load<u64>(ptr, 8) == 0x7900740069006E    // ytin
     ) {
-      return copysign<f64>(Infinity, sign);
+      return Infinity * sign;
```
Comment (dcodeIO) [inline std/assembly/util/string.ts:800]: What's the difference here? Just curious :)

File: std/assembly/number.ts (hunk 2 above)
Comment (MaxGraey) [reply std/assembly/number.ts:298]: also it reduce monomorphic variants to just `strtol<f64>` instead `strtol<f32>` and `strtol<f64>`

File: std/assembly/util/string.ts (hunk 5 above)
Comment (MaxGraey) [reply std/assembly/util/string.ts:800]: just smaller in term of codesize

Comment (dcodeIO) [reply std/assembly/util/string.ts:800]: Oh, is it? Would expect both binary ops having the same size in a binary.

Comment (MaxGraey) [reply std/assembly/util/string.ts:800]: Ah, yeah. Same size. But `mul` simpler for codegenerators.

Comment (MaxGraey) [reply std/assembly/util/string.ts:800]: https://godbolt.org/z/SkDktV

Comment (dcodeIO) [reply std/assembly/util/string.ts:800]: Do you know if there's a difference here on specific processor architectures? Would be surprised if `copysign` existed if it wasn't preferrable over `mul` in certain scenarios.

Comment (MaxGraey) [reply std/assembly/util/string.ts:800]: `copysign` always translate to several operations in all known archs. Also `mul` may be little slower but this case is really rare in any case

Comment (dcodeIO) [reply std/assembly/util/string.ts:800]: From another perspective: So, would it be feasible for Binaryen to optimize from `mul` to `copysign`?

Comment (MaxGraey) [reply std/assembly/util/string.ts:800]: I don't think so, it's not reduce codesize but require extra analysis and not always safe
